### DISCUSSION TITLE
MoodNft.sol - doubts on flipMood() conditions

### DISCUSSION THREAD

**Discussion by numacoding** [Cyfrin/foundry-full-course-cu#1176] [Q&A]
Hello everyone! Hope you all are having an amazing week. 
I'm working on the understanding of the `MoodNft` contract, and I saw the following if statement in the `flipMood()` function:

```
if (getApproved(tokenId) != msg.sender && ownerOf(tokenId) != msg.sender) {
            revert MoodNft__CantFlipMoodIfNotOwner();
        }
```

I try to code first and then see what Patrick did, and I used this condition:

```
if (ownerOf(_tokenId) != msg.sender) {
            revert MoodNft__NotNFTOwner();
        }
```

My question is about the first if statement. The logical operator shouldn't be OR instead of AND? Isn't the use of && the same that only having `ownerOf(tokenId) != msg.sender` as a condition? 
I'd appreciate the help to understand this issue. I also think that this might be a way of being as explicit as possible, but I think that, if the condition is redundant, we could save some gas by leaving just one condition.

Thank you, all!

**Comment by xyz899** [comment]
### Hi, Hope you're having a great day as well !,
#### In the contract Patrick does we want to make sure of two things :

1. First, we need to check that the owner that has granted permission (`getApproved`) to transfer the `tokenId` is the `msg.sender` itself, probably because the authorized address is here is the `msg.sender`, the one able to transfer the NFT.
2. Secondly, we want to make sure that the `msg.sender` is the one with owning the NFT indexed with the `tokenId`.

If we proceed to use || (OR) instead of && (AND), if someone malicious wanted to transfer the NFT to himself He would need to bypass only one of the two statements, so we make sure both condition which are referred on OpenZeppelin [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/cffb2f1ddcd87efd68effc92cfd336c5145acabd/contracts/token/ERC721/IERC721.sol#L127) are respected.

_winter99_

**Reply by numacoding** [reply]
Thank you, Winter! Got it. An OR statement would represent a vulnerability, thanks! Makes sense now. This condition will also prevent the case in which the owner gives permission to another EOA account or Smart Contract under the condition to transfer that NFT in a particular state, but suddenly changes it before the transaction.
Thank you for your help and for making me think!
